orange/src/main/java/com/hartwig/hmftools/orange/algo/immuno/ImmuneEscapeInterpreter.java

```diff
@@ -0,0 +1,162 @@
+import com.hartwig.hmftools.datamodel.purple.PurpleCodingEffect;
+import com.hartwig.hmftools.datamodel.purple.PurpleGainLoss;
+import com.hartwig.hmftools.datamodel.purple.PurpleGeneCopyNumber;
+import com.hartwig.hmftools.datamodel.purple.PurpleRecord;
+import com.hartwig.hmftools.datamodel.purple.PurpleVariant;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.jetbrains.annotations.NotNull;
+
+public final class ImmuneEscapeInterpreter
+{
+    private static final Logger LOGGER = LogManager.getLogger(ImmuneEscapeInterpreter.class);
+
+    static final Set<String> HLA_GENES = Sets.newHashSet("HLA-A", "HLA-B", "HLA-C");
+    static final Set<String> ANTIGEN_PRESENTATION_GENES = Sets.newHashSet("B2M", "CALR", "TAP1", "TAP2", "TABBP", "NLRC5", "CIITA", "RFX5");
+    static final Set<String> IFN_GAMMA_PATHWAY_GENES = Sets.newHashSet("JAK1", "JAK2", "IRF2", "IFNGR1", "IFNGR2", "APLNR", "STAT1");
+    static final Set<String> PD_L1_GENES = Sets.newHashSet("CD274");
+    static final Set<String> CD58_GENES = Sets.newHashSet("CD58");
+    static final Set<String> EPIGENETIC_SETDB1_GENES = Sets.newHashSet("SETDB1");
+
+    @NotNull
+    public static ImmuneEscapeRecord interpret(@NotNull PurpleRecord purple, @NotNull LinxRecord linx)
+    {
+        return ImmutableImmuneEscapeRecord.builder()
+                .hasHlaEscape(anyGeneWithLOH(purple, HLA_GENES) || anyGeneWithInactivation(purple, linx, HLA_GENES))
+                .hasAntigenPresentationPathwayEscape(anyGeneWithInactivation(purple, linx, ANTIGEN_PRESENTATION_GENES))
+                .hasIFNGammaPathwayEscape(anyGeneWithInactivation(purple, linx, IFN_GAMMA_PATHWAY_GENES))
+                .hasPDL1OverexpressionEscape(anyGeneWithAmplification(purple, PD_L1_GENES))
+                .hasCD58InactivationEscape(anyGeneWithInactivation(purple, linx, CD58_GENES))
+                .hasEpigeneticSETDB1Escape(anyGeneWithAmplification(purple, EPIGENETIC_SETDB1_GENES))
+                .build();
+    }
+
+    private static boolean anyGeneWithLOH(@NotNull PurpleRecord purple, @NotNull Set<String> genesToCheck)
+    {
+        for(String geneToCheck : genesToCheck)
+        {
+            if(hasLOH(purple.allSomaticGeneCopyNumbers(), geneToCheck))
+            {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    private static boolean hasLOH(@NotNull List<PurpleGeneCopyNumber> allSomaticGeneCopyNumbers, @NotNull String geneToCheck)
+    {
+        for(PurpleGeneCopyNumber somaticGeneCopyNumber : allSomaticGeneCopyNumbers)
+        {
+            if(somaticGeneCopyNumber.gene().equals(geneToCheck))
+            {
+                double minMajorAlleleCopyNumber = somaticGeneCopyNumber.minCopyNumber() - somaticGeneCopyNumber.minMinorAlleleCopyNumber();
+                return somaticGeneCopyNumber.minMinorAlleleCopyNumber() < 0.3 && minMajorAlleleCopyNumber > 0.7;
```

Comment:
Does the name minMajorAlleleCopyNumber actually make sense? Since the minCopyNumber and minMinorAlleleCopyNumber can be from different copy number regions, "minMajorAlleleCopyNumber" doesn't have to be the major allele copy number anywhere. For instance if the gene partially has CN==3 with minorAlleleCN==0 and partially CN==2 and minorAlleleCN==1, then this calculation gives 2, which isn't the major allele copy number anywhere.

Also, if we only call deletions when minCN<0.5, and can only call LOH when minCN>0.7+minMinorAlleleCN. Then if minMinorAlleleCN is 0 and minCN is 0.6, we call neither LOH nor a deletion, which is unfortunate since the only actual unclear bit in that situation is which of the two it is. And if minMinorAlleleCN is greater than 0, then the minCN range where we can call neither deletion nor LOH increases. I think it isn't great that we would miss the inactivation in that case.

I think we should consider replacing the minMajorAlleleCopyNumber part of the check with `somaticGeneCopyNumber.minCopyNumber() >= 0.5`. 

The only counterargument to this I can think of is that perhaps in that in between region we think it's a subclonal deletion, and that we don't want to call that as escape since it isn't clonal. But it doesn't sound like a great argument to me, and it definitely doesn't apply when minMinorAlleleCN is exactly 0.

Reply:
Good suggestion. I've tried to stay as close to the vignette implementation as possible, but didn't feel exactly right. This is better, and will let Joseph know I've changed this. 

Did make it ">", I am not a huge fan of using = for comparing doubles, check for LOH now is:

`somaticGeneCopyNumber.minCopyNumber() > 0.5  && somaticGeneCopyNumber.minMinorAlleleCopyNumber() < 0.3`